drivers/i2c/meson/driver.h

```diff
@@ -44,11 +44,18 @@ typedef struct _i2c_ifState {
     enum data_direction data_direction;
     /* I2C bus address of the current request being handled */
     size_t addr;
+
+    unsigned long timeout;
 } i2c_ifState_t;
 
 #define DATA_DIRECTION_WRITE (0x0)
 #define DATA_DIRECTION_READ (0x1)
 
+// Max number of timeout IRQs before giving up. For some devices such as the pn532,
+// these arrive constantly in normal operation. As a result we need a high threshold
+// before acting.
+#define TIMEOUT_MAX 100000
```

Comment:
Seems weird to work-around broken hardware in the driver rather than the client.

Reply:
The real issue is that we don't know why the timeouts get generated. The main thing we need this driver to do is to avoid locking up in the case that an actual defective device ACKs and then stops. Since the I2C hardware doesn't stop when an IRQ arrives, it's really hard to reason about whether any particular timeout is authentic, so I figured that just counting them is an OK fix since we know that a large quantity continuously discounts the possibility of a weird device.

Since we have no way to know why the IRQs fire, we can't really figure out why the PN532 causes it. IMO it's not worth leaving this driver in a state where it can lock up because of that

Reply:
(as in the timeout IRQs are undocumented - the actual condition that makes them fire is unclear to me even when looking at the logic analyser output)